src/worktree.rs

```diff
@@ -0,0 +1,331 @@
+    Locked(Option<String>),
+}
+
+impl Worktree {
+    /// Open a worktree of a the repository
+    ///
+    /// If a repository is not the main tree but a worktree, this
+    /// function will look up the worktree inside the parent
+    /// repository and create a new `git_worktree` structure.
+    pub fn open_from_repository(repo: &Repository) -> Result<Worktree, Error> {
+        let mut raw = ptr::null_mut();
+        unsafe {
+            try_call!(raw::git_worktree_open_from_repository(&mut raw, repo.raw()));
+            Ok(Binding::from_raw(raw))
+        }
+    }
+
+    /// Retrieves the name of the worktree
+    ///
+    /// This is the name that can be passed to repo::Repository::worktree_lookup
+    /// to reopen the worktree. This is also the name that would appear in the
+    /// list returned by repo::Repository::worktrees
+    pub fn name(&self) -> Option<&str> {
+        unsafe {
+            crate::opt_bytes(self, raw::git_worktree_name(self.raw))
+                .and_then(|s| str::from_utf8(s).ok())
+        }
+    }
+
+    /// Retrieves the path to the worktree
+    ///
+    /// This is the path to the top-level of the source and not the path to the
+    /// .git file within the worktree. This path can be passed to
+    /// repo::Repository::open.
+    pub fn path(&self) -> &Path {
+        unsafe {
+            util::bytes2path(crate::opt_bytes(self, raw::git_worktree_path(self.raw)).unwrap())
+        }
+    }
+
+    /// Validates the worktree
+    ///
+    /// This checks that it still exists on the
+    /// filesystem and that the metadata is correct
+    pub fn validate(&self) -> Result<(), Error> {
+        unsafe {
+            call::c_try(raw::git_worktree_validate(call::convert(&self.raw)))?;
+        }
+        Ok(())
+    }
+
+    /// Locks the worktree
+    pub fn lock(&self, reason: Option<&str>) -> Result<(), Error> {
+        let reason = crate::opt_cstr(reason)?;
+        unsafe {
+            try_call!(raw::git_worktree_lock(self.raw, reason));
+        }
+        Ok(())
+    }
+
+    /// Unlocks the worktree
+    pub fn unlock(&self) -> Result<(), Error> {
+        unsafe {
+            call::c_try(raw::git_worktree_unlock(call::convert(&self.raw)))?;
+        }
+        Ok(())
+    }
+
+    /// Checks if worktree is locked
+    pub fn is_locked(&self) -> Result<WorktreeLockStatus, Error> {
+        let buf = Buf::new();
+        unsafe {
+            match try_call!(raw::git_worktree_is_locked(buf.raw(), self.raw)) {
+                0 => Ok(WorktreeLockStatus::Unlocked),
+                _ => {
+                    println!("Buf: {}", buf.as_str().unwrap());
+                    let v = buf.to_vec();
+                    println!("Length of v: {}", v.len());
+                    Ok(WorktreeLockStatus::Locked(match v.len() {
+                        0 => None,
+                        _ => String::from_utf8(v).ok(),
+                    }))
+                }
```

Comment:
`git_worktree_is_locked` can return <0 when error and >0 for the locked state, which I think should be handled separately. Also, the `println`s are still there ;)

Reply:
`try_call!` macro handles when the return value is `< 0` by calling `c_try`